egg-herbie/main.rkt

```diff
@@ -60,6 +60,92 @@
 ; Checks for a condition on MacOS if x86 Racket is being used on an ARM mac.
 (define-ffi-definer define-eggmath (ffi-lib libeggmath-path #:fail handle-eggmath-import-failure))
 
+;; Frees a Rust-allocated C-string
+(define-eggmath destroy_string (_fun _pointer -> _void))
+
+;; Gets the length of a Rust-allocated C-string in bytes,
+;; excluding the nul terminator.
+(define-eggmath string_length (_fun _pointer -> _uint32))
+
+;; Converts a Racket string to a C-style string.
+(define (string->_rust/string s #:raw? [raw? #f])
+  (define bstr (string->bytes/utf-8 s))
+  (define n (bytes-length bstr))
+  (define p (malloc (if raw? 'raw 'atomic) (add1 n)))
+  (memcpy p bstr n)
+  (ptr-set! p _byte n 0)
+  p)
+
+;; Converts a non-NULL, Rust-allocated C-string to a Racket string,
+;; freeing the Rust string.
+(define (_rust/string->string p)
+  (define len (string_length p))
+  (define bstr (make-bytes len))
+  (memcpy bstr p len)
+  (destroy_string p)
+  (bytes->string/utf-8 bstr))
+
+;; Converts a non-NULL, Rust-allocated C-string to a Racket datum
+;; by repeatedly reading the string. The underlying Rust string
+;; is automatically freed.
+(define (_rust/string->data p)
+  (define len (string_length p))
+  (define bstr (make-bytes len))
+  (memcpy bstr p len)
+  (destroy_string p)
+  (for/list ([datum (in-port read (open-input-bytes bstr))])
+    datum))
+
+;; FFI type that converts Rust-allocated C-style strings
+;; to Racket strings, automatically freeing the Rust-side allocation.
+(define _rust/string
+  (make-ctype _pointer
+              (lambda (x) (and x (string->_rust/string x)))
+              (lambda (x) (and x (_rust/string->string x)))))
+
+;; FFI type that converts Rust-allocated C-style strings
+;; to a Racket datum via `read`, automatically freeing the Rust-side allocation.
+(define _rust/datum
+  (make-ctype _pointer
+              (lambda (x) (and x (string->_rust/string (~a x))))
+              (lambda (x) (and x (first (_rust/string->data x))))))
+
+;; FFI type that converts Rust-allocated C-style strings
+;; to multiple Racket datum via reapeated use of `read`,
+;; automatically freeing the Rust-side allocation.
+(define _rust/data
+  (make-ctype _pointer
+              (lambda (_) (error '_rust/data "cannot be used as an input type"))
+              (lambda (x) (and x (_rust/string->data x)))))
+
+; Egraph iteration data
+; Not managed by Racket GC.
+; Must call `destroy_egraphiters` to free.
+(define-cstruct _EGraphIter ([numnodes _uint] [numeclasses _uint] [time _double]) #:malloc-mode 'raw)
```

Comment:
I think there's a way to register destructors for C structs. Not a big deal.

Reply:
Yeah, with `register-finalizer`

Reply:
These are manually freed but they're no longer exported, so I figured it's not a huge deal

Reply:
Sure